GH-9492: Fix @SpringIntegration in @Nested test

mtomik opened the pull request:
Hi,
this Pull Request aims to fix issue with using @SpringIntegration test annotation in conjuction with @Nested classes
more information can be found here  #9492

mtomik committed:
GH-9492: Fix @SpringIntegration in @Nested test

pivotal-cla commented:
@mtomik Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration&pullRequestId=9506)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-integration&pullRequestId=9506) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@mtomik Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration&pullRequestId=9506)!

artembilan commented on spring-integration-test/src/main/java/org/springframework/integration/test/context/SpringIntegrationTestExecutionListener.java:
```diff
@@ -66,4 +75,15 @@ private boolean match(String name, String[] patterns) {
+	/**
+	 * to allow nested class execution
+	 */
+	private MockIntegrationContext handledGetMockIntegrationContext(ApplicationContext applicationContext) {
```
See the `MockIntegrationContextCustomizerFactory` for similar `TestContextAnnotationUtils` fix instead.

onobc replied:
Got it [here](https://github.com/spring-projects/spring-integration/pull/9506/files#diff-7c2fcf66d0bd0e26c5966321b6dd529e60067a1d1c36200d3ca82fdabbeb6e47R42)

artembilan reviewed (changes requested):
Thank you for looking into this! The problem is really not  so easy to spot.
We ended up with @onobc for the same conclusion as you did.

We will probably ask @sbrannen to improve docs for the `@NestedTestCinfiguration` to mention that this `TestContextAnnotationUtils` is important to use in custom annotation processors if we’d like make them working with `@Nested` as well.

We still would like to see some test suite to cover this functionality and supplement the fix.

You are right, something like AMQP cannot be used in this module, but that still is out of the annotation in the subject responsibility. The listener container you are mentioning in your test is definitely not an `AbstractEndpoint` we worry here about.

onobc committed:
GH-9492: Add tests and extend fix to MockIntegrationContext

Switches `MockIntegrationContextCustomizerFactory` to use
`TestContextAnnotationUtils` in order to properly support
`@NestedTestConfiguration` semantics.

Adds an integration test with a base class and several
`@Nested` inner test classes to verify the various
permutations of `@SpringIntegrationTest` inherit/override
behavior when used w/ `@NestedTestConfiguration`.

onobc commented:
> Thank you for looking into this! The problem is really not so easy to spot. We ended up with @onobc for the same conclusion as you did.

I agree w/ @artembilan in that It was not easy to spot and you did a great job @mtomik!
 
> We still would like to see some test suite to cover this functionality and supplement the fix.

I went ahead and contributed this in commit 2. Me and @artembilan paired this afternoon and had the beginnings of the tests so I went ahead and wrapped them up and added to this code proposal. I verified the tests fail as expected w/o the fixes.

> 
> You are right, something like AMQP cannot be used in this module, but that still is out of the annotation in the subject responsibility. The listener container you are mentioning in your test is definitely not an `AbstractEndpoint` we worry here about.

sbrannen commented:
> We will probably ask @sbrannen to improve docs for the `@NestedTestCinfiguration` to mention that this `TestContextAnnotationUtils` is important to use in custom annotation processors if we’d like make them working with `@Nested` as well.

I've created the following issue to address that.

- https://github.com/spring-projects/spring-framework/issues/33586

mtomik commented:
Thanks guys for quick feedback :+1:

artembilan commented on spring-integration-test/src/test/java/org/springframework/integration/test/context/AbstractIntegrationTest.java:
```diff
@@ -0,0 +1,50 @@
+ * to be inherited by concrete subclasses.
+ *
+ * @author Chris Bono
+ * @since 6.4.0
```
I think we can treat this as a bug, plus this is just a testing feature, so I'm going to back port the fix down to `6.2.x`.
Therefore `@since 6.2.10` over here.

artembilan commented on spring-integration-test/src/test/java/org/springframework/integration/test/context/NestedSpringIntegrationTestAnnotationTests.java:
```diff
@@ -0,0 +1,86 @@
+ * when used with {@link Nested} and {@link NestedTestConfiguration}.
+ *
+ * @author Chris Bono
+ * @since 6.4.0
```
DITTO

artembilan commented on spring-integration-test/src/test/java/org/springframework/integration/test/context/NestedSpringIntegrationTestAnnotationTests.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	@Nested
+	@NestedTestConfiguration(NestedTestConfiguration.EnclosingConfiguration.INHERIT)
```
I don't think we need this specific test since `INHERIT` is there by default.

artembilan commented on spring-integration-test/src/test/java/org/springframework/integration/test/context/NestedSpringIntegrationTestAnnotationTests.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	@Nested
+	@NestedTestConfiguration(NestedTestConfiguration.EnclosingConfiguration.INHERIT)
```
This line is redundant since `INHERIT` is by default.

artembilan reviewed (changes requested):
Thank you all for contribution!
Not clear from now how is going to fix the rest of review concerns, but I can do that myself on merge to make all our lives easier :smile.

artembilan reviewed (approved):
I'm taking this PR locally for final touches.

Thank you all for contribution!

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/3a8e3abc7d2f25fe39290a9758e9d556cc0faf22.
And back-ported to `6.3.x` & `6.2.x` respectively.
Thank you again for the contribution!
